src/ballet/block/fd_microblock.h

```diff
@@ -0,0 +1,180 @@
+
+FD_PROTOTYPES_BEGIN
+
+/* fd_microblock_{align,footprint} return the required alignment and
+   footprint of a memory region suitable for storing an `fd_microblock_t`.
+   `fd_microblock_align` returns `fd_microblock_ALIGN`.
+   `fd_microblock_footprint` returns the memory footprint of a `fd_microblock_t`
+   that can store a given number of maximally-sized txns. */
+
+FD_FN_CONST ulong
+fd_microblock_align( void );
+
+FD_FN_CONST ulong
+fd_microblock_footprint( ulong txn_max_cnt );
+
+/* fd_microblock_new formats an unused memory region for storing an
+   `fd_microblock_t`. `shmem` is a non-NULL pointer to this region in the
+   local address space with the required footprint and alignment.
+   `txn_max_cnt` is the number of txn descriptors that can be stored
+   in the `fd_microblock_t`.  The `fd_microblock_hdr_t` is initialized to zero.
+   Returns `shmem` on success and NULL on failure (logs details).
+   Reasons for failure include an obviously bad `shmem` (alignment)
+   or bad `txn_max_cnt`. */
+
+void *
+fd_microblock_new( void * shmem,
+                   ulong  txn_max_cnt );
+
+/* `fd_microblock_join` joins the caller to the microblock.
+   `shblock` points to the first byte of the memory region backing the
+   `fd_microblock_t` in the caller's address space.
+   Returns `shblock` on success and NULL on failure (logs details).
+   Reasons for failure include those of `fd_microblock_new` or that the
+   memory region was not correctly initialized.
+   Every successful join should have a matching leave. The lifetime of
+   the join is until the matching leave or thread group is terminated. */
+
+fd_microblock_t *
+fd_microblock_join( void * shblock );
+
+/* `fd_microblock_leave` leaves a current local join.  Returns a pointer to
+   the underlying shared memory region on success.  Reasons for failure
+   include `block` is NULL. */
+
+void *
+fd_microblock_leave( fd_microblock_t * block );
+
+/* `fd_microblock_delete` unformats a memory region storing an `fd_microblock_t`.
+   Assumes nobody is joined to the region.  Returns a pointer to the
+   underlying shared memory region or NULL if used obviously in error
+   (e.g. `shblock` is obviously not an `fd_microblock_t` ... logs details).
+   The ownership of the memory region is transferred to the caller. */
+
+void *
+fd_microblock_delete( void * shblock );
+
+/* fd_microblock_deserialize: Deserializes a microblock from `buf`.
+   Only accesses up to `buf_sz` bytes beyond `buf`.
+
+   Returns 1 on success and 0 on failure.
+   Reasons for failure include invalid data or unexpected EOF.
+
+   If the returned ptr is non-NULL, it is guaranteed to be
+     1) greater than the provided `buf` ptr
+     2) less or equal than `(uchar *)buf + buf_sz` */
+int
+fd_microblock_deserialize( fd_microblock_t * block,
+                           void **           buf,
+                           ulong *           buf_sz,
+                           fd_txn_parse_counters_t * counters_opt );
```

Comment:
I was confused initially because it doesn't return a pointer. I haven't read the implementation yet, but my guess is `buf` is an in/out parameter to facilitate parsing multiple microblocks from a single buffer.

What about taking `buf` as a `uchar const *` and returning a `ulong` which is just how much data was consumed or 0 on error. I could be persuaded to switch to `void *`. Either way, it should be `const`, and I much prefer returning a value instead of using output parameter unless there's a strong reason for it.